### PR TITLE
BBA/BuiltIn: Add UPnP HTTP listener

### PR DESCRIPTION
**Pull request by sepalani**
This PR implements the UPnP HTTP listener used by Kirby Air Ride and Mario Kart Double Dash. If someone on the same network of these GC games visits the URL `http://<GameCube_IP>:1900/rootDesc.xml`, they can grab the UPnP XML file. I confirmed this behaviour using Nintendont on a real Wii (since I don't have a BBA adapter). **This URL can only be accessed from another IP (i.e. an IP other than the one used by Dolphin).**

@schthack @nolrinale 
This might also address this issue https://dolp.in/i12996 as this PR fixes an issue where partial data isn't sent when closing the connection. I touched the TCP emulation part so you might want to test this out to make sure it doesn't impact PSO servers.

This PR doesn't fix MKDD timer stuck at 172 when there are more than 2 players. Depends on https://github.com/dolphin-emu/dolphin/pull/10920.

### CODE REVIEW

**Comment by Notaloop763** [comment]
Would it be possible to create automated testing like what fifo does for graphics? Each change affects several games and has the chance to introduce regressions.

**Comment by sepalani** [comment]
> Would it be possible to create automated testing like what fifoCI does for graphics? Each change affects several games and has the chance to introduce regressions.

This would be quite complicated. We already have UnitTest test cases to ensure some features of Dolphin are working properly. I could write hardware tests (homebrew) to check some networking behaviours like I did for small Wii networking edge cases. The main issue with complex cases like this one would be the reliability. Timing might alter the test results, I don't think TASing would work since the network behaviour isn't very deterministic.

Overall, I don't think that's easily feasible as of yet.

**Comment by nolrinale** [comment]
I've tested the PR and here are the findings from my side:

-The `Warning: Partial sends might not be handled properly.` error continues as usual.

-By adding a breakpoint into the `311` line in `BuiltIn.cpp` file I was able to stop the operation at this point in the code by just accessing the Dolphin host computer ip address from another device (My iPad), every time I tried to just type `mymacip:1900` from the iPad Safari, the breakpoint would be triggered on the main Mac, I didn't even needed to append the `/rootDesc.xml` (which is something you might need to further control). I also did tested appending the whole URL like in your example but the xml file didn't generated and auto downloaded in the Safari browser of the iPad either (but the breakpoint still got triggered tho).

I might not be too savvy with networking security but I think its a security issue opening a port and setting it up to receive/send crafted data like this xml like a pseudo small web server service. Maybe there's another way without resorting to this method?

**Comment by sepalani** [comment]
@nolrinale
I think you missed the point of this PR which is allowing to access the UPnP HTTP listener that is used by GC games using SSDP (MKDD, Kirby). The way the GC does it is by opening the port 1900 and listens for incoming HTTP requests. What this PR is doing is opening said port and forwards the requests to the emulated GameCube via the BBA/BuiltIn code. AFAICT, PSO doesn't seem to use SSDP so it doesn't have a listener on port 1900 and visiting this URL shouldn't produce anything meaningful. The Dolphin code doesn't generate these XML files but the game does it.

Opening a port shouldn't be a security issue per se, even if it does increase the network attack surface. Unless the game is vulnerable to remote code execution attacks from its network stack, this shouldn't be an issue here as we only forward the network traffic to the BBA code. If such vulnerability exists, then you found a RCE for MKDD and Kirby Air Ride (which would be nice for homebrew related stuff on non-modded GameCube).

If this doesn't fix the PSO warning, then the error is located somewhere else. I just want to make sure this PR (which improve the BBA emulation accuracy for other games) doesn't break existing features as I don't own all BBA games. Importing items in my country is a pain and not cheap either, if I find the opportunity to grab some of these games I will but that's clearly not my priority, at the moment.

**Comment by Notaloop763** [comment]
@sepalani  I can test this build today on windows 10 x64 for a few hours.

Is there something specific I should do to test your changes? Like, run pr-win-x64 build with specific settings or running the pr-win-dbg-x64 build?

**Comment by Notaloop763** [comment]
I"ve been testing the pr-win-x64 build for a few hours and haven't disconnected. Seems to be stable for my build.

**Comment by JMC47** [comment]
Everything seems to be working as expected, but does need that other PR.  I think we just need a review on these.

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.h [hunk 1]
```diff
@@ -41,6 +42,8 @@ class BbaTcpSocket : public sf::TcpSocket
   BbaTcpSocket();
 
   sf::Socket::Status Connect(const sf::IpAddress& dest, u16 port, u32 net_ip);
+  sf::Socket::Status GetPeerName(sockaddr_in* addr);
```
Comment (iwubcode) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.h:45]: Can these be `const`?

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp [hunk 2]
```diff
@@ -727,6 +779,22 @@ sf::Socket::Status BbaTcpSocket::Connect(const sf::IpAddress& dest, u16 port, u3
   return this->connect(dest, port);
 }
 
+sf::Socket::Status BbaTcpSocket::GetPeerName(sockaddr_in* addr)
```
Comment (iwubcode) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:782]: This isn't a lot of code...but maybe add a helper function that both implementations call?

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp [hunk 3]
```diff
@@ -669,6 +685,37 @@ void CEXIETHERNET::BuiltInBBAInterface::ReadThreadHandler(CEXIETHERNET::BuiltInB
       }
     }
 
+    // Check for new UPnP client
```
Comment (iwubcode) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:688]: Could this be moved to a new function?  Just my take bu this function seems like it's doing a lot..

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp [hunk 4]
```diff
@@ -669,6 +685,37 @@ void CEXIETHERNET::BuiltInBBAInterface::ReadThreadHandler(CEXIETHERNET::BuiltInB
...
+      for (auto& tcp_buf : ref->tcp_buffers)
+        tcp_buf.used = false;
+      ref->bba_mac = self->m_current_mac;
+      ref->my_mac = self->ResolveAddress(ref->from.sin_addr.s_addr);
+      ref->tcp_socket.setBlocking(false);
+      ref->ready = false;
+
+      Common::TCPPacket result(ref->bba_mac, ref->my_mac, ref->from, ref->to, ref->seq_num,
+                               ref->ack_num, TCP_FLAG_SIN);
+      result.tcp_options = {0x02, 0x04, 0x05, 0xb4, 0x01, 0x03, 0x03, 0x08, 0x01, 0x01, 0x04, 0x02};
```
Comment (iwubcode) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:715]: What is this?  Maybe a comment to explain?

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp [hunk 5]
```diff
@@ -678,7 +725,12 @@ void CEXIETHERNET::BuiltInBBAInterface::ReadThreadHandler(CEXIETHERNET::BuiltInB
       {
         SetIPIdentification(buffer, datasize, ++self->m_ip_frame_id);
       }
-      self->m_eth_ref->mRecvBufferLength = datasize > 64 ? static_cast<u32>(datasize) : 64;
+      if (datasize < 64)
+      {
+        std::fill(buffer + datasize, buffer + 64, 0);
```
Comment (iwubcode) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:730]: How do we know that the buffer is large enough?

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp [hunk 6]
```diff
@@ -298,18 +302,30 @@ void CEXIETHERNET::BuiltInBBAInterface::HandleTCPFrame(const Common::TCPPacket&
...
-  const u16 properties = ntohs(tcp_header.properties);
-  if (properties & (TCP_FLAG_FIN | TCP_FLAG_RST))
+  const u16 flags = ntohs(tcp_header.properties) & 0xfff;
+  if (flags & (TCP_FLAG_FIN | TCP_FLAG_RST))
   {
     if (ref == nullptr)
       return;  // not found
 
-    ref->ack_num++;
+    ref->ack_num += 1 + static_cast<u32>(data.size());
```
Comment (iwubcode) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:311]: All the other `ack_num` values being set use `ntohl` did you intentionally leave it off here for some reason?

Comment (sepalani) [reply Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:311]: `ack_num` is using the host's endianness and is used to acknowledge the TCP packet bytes (sequence number) received. When inspecting TCP packets from the BBA buffer we need to convert from _network to host_ (`ntoh*`) endianness. Here we're just incrementing it (not setting it from a network packet) so we don't need to swap it.

**Comment by sepalani** [comment]
@iwubcode 
Done.

I suspect Dolphin's HLE BBA to have similar limitations with [Nintendont HLE BBA](https://github.com/FIX94/Nintendont/blob/master/BBA_Readme.md). I wonder if it's related to the MAC addresses not being valid BBA ones.

**Comment by sepalani** [comment]
After some testing, it seems to behave just like before the rebase. I also found the issue I had with one of my PC and its VM:
 - https://communities.vmware.com/t5/VMware-Workstation-Pro/SSDP-blocker/td-p/2305646

**Comment by sepalani** [comment]
I inverted the first `if` in `HandleUPnPClient` to return early and go down from one indentation level. Otherwise, there is no functional difference with the commit I amended.

**Comment by JMC47** [comment]
I'm surprised this doesn't conflict with 10920?  Is this good to merge as well?

**Comment by sepalani** [comment]
> I'm surprised this doesn't conflict with 10920? Is this good to merge as well?

I'm surprised too, I did a rebase just for good measure. Though, I'm surprised that GitHub was still showing 4 commits to be merged instead of one.

I'm also done with this PR.
